Sources/NSScanner+Swift.swift

```diff
@@ -36,6 +36,20 @@ extension Scanner {
         return nil
     }
     
+    /// Returns the given string if scanned, or `nil` if not found.
+    @discardableResult func scanString(_ str: String, options: NSString.CompareOptions) -> String? {
+        let scanRange = NSRange(location: self.scanLocation, length: str.count)
+        guard scanRange.location + str.count <= self.string.count else { return nil }
+
+        let scanned = (self.string as NSString).substring(with: scanRange)
+        if (scanned.compare(str, options: options, range: nil, locale: nil) == .orderedSame) {
+            self.scanLocation += str.count
+            return str
+        } else {
+            return nil
+        }
+    }
```

Comment:
- NSString is based UTF-16
- String.count is calculated by UTF-8.

NSString.length vs String.count
```swift
"👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦".count // 4
("👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦" as NSString).length // 44
```

Scanner is made of NSScanner, it basically operates as an NSString.
Therefore, scanLocation is estimated based on UTF-16.

`self.string.count` has a problem when using compatibility string on your code.


```suggestion
    @discardableResult func scanString(_ str: NSString, options: NSString.CompareOptions) -> String? {
        let scanRange = NSRange(location: self.scanLocation, length: str.length)
        let string = self.string as NSString
        guard scanRange.location + str.length <= string.length else { return nil }

        let scanned = string.substring(with: scanRange)
        if (scanned.compare(str as String, options: options, range: nil, locale: nil) == .orderedSame) {
            self.scanLocation += str.length
            return str as String
        } else {
            return nil
        }
    }
```

Test code
```swift
// sut should not encounter infinity loop when using compatibility string
func testCompatibilityString() {
    XCTAssertEqual("<font>Hello 👨‍👧‍👧</font>".detectTags().string, "Hello 👨‍👧‍👧")
    XCTAssertEqual("<font>Hello 👨‍👧‍👧👨‍👧‍👧</font>".detectTags().string, "Hello 👨‍👧‍👧👨‍👧‍👧")
    XCTAssertEqual("<font>Hello 👨‍👧‍👧👨‍👧‍👧👨‍👧‍👧</font>".detectTags().string, "Hello 👨‍👧‍👧👨‍👧‍👧👨‍👧‍👧")
    XCTAssertEqual("<font>Hello 👨‍👧‍👧👨‍👧‍👧👨‍👧‍👧👨‍👧‍👧</font>".detectTags().string, "Hello 👨‍👧‍👧👨‍👧‍👧👨‍👧‍👧👨‍👧‍👧")
    XCTAssertEqual("<font>Hello 👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦</font>".detectTags().string, "Hello 👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦👨‍👨‍👦‍👦")
}
```